### PR TITLE
An very explosive port and an minelayer update

### PR DESCRIPTION
**Pull request by thedarkcrusader**
## About The Pull Request
Ported deployable mines from beestation they will only blow up when they are deployed they can also be disarmed and picked back up i also fixed the sharp mine being available for everyone to craft also added a new recipe the explosive mine which also can only be unlocked with the mine layer blueprint

## Why It's Good For The Game
Now you can carry your mines  and wont have to be fucked when that one guy crafts a mine right next to your matrix point also because i think its better then what we currently have.

## Changelog
:cl:
add: ported deployable mines from beestation
add: added new explosive mine recipe available via the minelayer blueprint 
s
balance: the sharpnel mine recipe can no longer be crafted by everyone
/:cl:

### CODE REVIEW

File: code/datums/components/crafting/recipes/recipes_weapon_and_ammo.dm [hunk 1]
```diff
@@ -139,15 +139,30 @@
...
+	always_availible = FALSE
+
+
+/datum/crafting_recipe/explosivemine
+	name = "explosive Mine"
+	result = /obj/item/deployablemine/explosive
+	reqs = list(/obj/item/grenade/homemade/firebomb = 1,
+				/obj/item/stack/tile = 1,
+				/obj/item/stack/ore/blackpowder     = 3
+				)
```
Comment (Rohesie) [inline code/datums/components/crafting/recipes/recipes_weapon_and_ammo.dm:160]:
```suggestion
	reqs = list(
		/obj/item/grenade/homemade/firebomb = 1,
		/obj/item/stack/tile = 1,
		/obj/item/stack/ore/blackpowder = 3,
		)
```
* Properly format lists.
* Make use of trailing commas.

File: code/datums/components/crafting/recipes/recipes_weapon_and_ammo.dm [hunk 2]
```diff
@@ -139,15 +139,30 @@
...
+	time = 15
```
Comment (Rohesie) [inline code/datums/components/crafting/recipes/recipes_weapon_and_ammo.dm:161]:
```suggestion
	time = 1.5 SECONDS
```
* Make use of time defines.

File: code/game/objects/effects/mines.dm [hunk 3]
```diff
@@ -1,25 +1,117 @@
+/obj/item/deployablemine
+	name = "deployable mine"
+	desc = "An unarmed landmine. It can be planted to arm it."
+	icon_state = "uglymine"
+	var/mine_type = /obj/effect/mine
+	var/arming_time = 30
```
Comment (Rohesie) [inline code/game/objects/effects/mines.dm:6]:
```suggestion
	var/arming_time = 3 SECONDS
```

File: code/game/objects/effects/mines.dm [hunk 4]
```diff
@@ -1,25 +1,117 @@
...
+/obj/item/deployablemine/smartstun
+	name = "deployable smart mine"
+	desc = "An unarmed smart stun mine. It can be planted to arm it."
+	mine_type = /obj/effect/mine/stun/smart
+
+/obj/item/deployablemine/rapid
+	name = "deployable rapid smart mine"
+	desc = "An unarmed smart stun mine designed to be rapidly placeable."
+	mine_type = /obj/effect/mine/stun/smart/adv
+	arming_time = 10
```
Comment (Rohesie) [inline code/game/objects/effects/mines.dm:21]:
```suggestion
	arming_time = 1 SECONDS
```

File: code/game/objects/effects/mines.dm [hunk 5]
```diff
@@ -1,25 +1,117 @@
...
+	w_class = WEIGHT_CLASS_SMALL
+
+/obj/item/deployablemine/heavy
+	name = "deployable sledgehammer smart mine"
+	desc = "An unarmed smart heavy stun mine designed to be hard to disarm."
+	mine_type = /obj/effect/mine/stun/smart/heavy
+	arming_time = 50
```
Comment (Rohesie) [inline code/game/objects/effects/mines.dm:28]:
```suggestion
	arming_time = 5 SECONDS
```

File: code/game/objects/effects/mines.dm [hunk 6]
```diff
@@ -1,25 +1,117 @@
...
+
+/obj/item/deployablemine/shrapnel
+	name = "deployable Sharpnel  mine"
+	desc = "An unarmed sharpnel  mine designed to spray lead everywhere and  be hard to disarm."
+	mine_type = /obj/effect/mine/shrapnel
+	arming_time = 50
```
Comment (Rohesie) [inline code/game/objects/effects/mines.dm:34]:
```suggestion
	arming_time = 5 SECONDS
```

File: code/game/objects/effects/mines.dm [hunk 7]
```diff
@@ -1,25 +1,117 @@
...
+	mine_type = /obj/effect/mine/gas/plasma
+
+/obj/item/deployablemine/sleepy
+	name = "knockout mine"
+	desc = "An unarmed mine that releases N2O into the air when triggered. Nighty Night!"
+	mine_type = /obj/effect/mine/gas/n2o
+
+/obj/item/deployablemine/afterattack(atom/plantspot, mob/user, proximity)
+	if(!proximity)
+		return
```
Comment (Rohesie) [inline code/game/objects/effects/mines.dm:57]: This proc sends a signal at the base of it, so make the redefinition call parent.

File: code/game/objects/effects/mines.dm [hunk 8]
```diff
@@ -1,25 +1,117 @@
...
+
+	if(!istype(plantspot,/turf/open)) // you can't plant a mine inside a wall or on a mob
+		return
```
Comment (Rohesie) [inline code/game/objects/effects/mines.dm:60]: Perhaps a turf property would be better.
You could add a flag at the turf type level.

File: code/game/objects/effects/mines.dm [hunk 9]
```diff
@@ -1,25 +1,117 @@
...
+	if((istype(plantspot,/turf/open/lava)) || (istype(plantspot,/turf/open/chasm)))
+		to_chat(user, "<span class='warning'>You can't plant the mine here!</span>")
+		return
+
+	to_chat(user, "<span class='notice'>You start arming the [src]...</span>")
+	if(do_after(user, arming_time, target = src))
+		new mine_type(plantspot)
+		to_chat(user, "<span class='notice'>You plant and arm the [src].</span>")
+		log_combat(user, src, "planted and armed")
+		qdel(src)
```
Comment (Rohesie) [inline code/game/objects/effects/mines.dm:75]:
```suggestion
	if(!do_after(user, arming_time, target = src))
		return
	new mine_type(plantspot)
	to_chat(user, "<span class='notice'>You plant and arm the [src].</span>")
	log_combat(user, src, "planted and armed")
	qdel(src)
```
Make use of early returns.
You might want to check earlier if there's a `mine_type` defined. After all one could conceive non-functional mines existing.

File: code/game/objects/effects/mines.dm [hunk 10]
```diff
@@ -1,25 +1,117 @@
...
 	name = "dummy mine"
 	desc = "Better stay away from that thing."
 	density = FALSE
 	anchored = TRUE
 	icon = 'icons/obj/items_and_weapons.dmi'
 	icon_state = "uglymine"
-	/// We manually check to see if we've been triggered in case multiple atoms cross us in the time between the mine being triggered and it actually deleting, to avoid a race condition with multiple detonations
 	var/triggered = FALSE
+	var/smartmine = 0
+	var/disarm_time = 200
```
Comment (Rohesie) [inline code/game/objects/effects/mines.dm:86]:
```suggestion
	var/disarm_time = 20 SECONDS
```

File: code/game/objects/effects/mines.dm (hunk 9 above)
Comment (thedarkcrusader) [reply code/game/objects/effects/mines.dm:75]: mine types should allready be defiend yep

File: code/game/objects/effects/mines.dm [hunk 11]
```diff
@@ -1,25 +1,117 @@
...
 /obj/effect/mine
 	name = "dummy mine"
 	desc = "Better stay away from that thing."
 	density = FALSE
 	anchored = TRUE
 	icon = 'icons/obj/items_and_weapons.dmi'
 	icon_state = "uglymine"
-	/// We manually check to see if we've been triggered in case multiple atoms cross us in the time between the mine being triggered and it actually deleting, to avoid a race condition with multiple detonations
 	var/triggered = FALSE
+	var/smartmine = 0
```
Comment (Rohesie) [inline code/game/objects/effects/mines.dm:85]: What is this? It looks like a boolean, but it's using numeric definitions.
Also, avoid `nospacescase` please, favor `snake_case`

File: code/game/objects/effects/mines.dm [hunk 12]
```diff
@@ -1,25 +1,117 @@
...
+	var/disarm_time = 200
+	var/disarm_product = /obj/item/deployablemine // ie what drops when the mine is disarmed
+
+/obj/effect/mine/attackby(obj/I, mob/user, params)
+	if(istype(I, /obj/item/multitool))
```
Comment (Rohesie) [inline code/game/objects/effects/mines.dm:90]: This proc sends a signal at the base. Make it call parent and take the return into account.

File: code/game/objects/effects/mines.dm (hunk 11 above)
Comment (thedarkcrusader) [reply code/game/objects/effects/mines.dm:86]: I dont know what this is its been a few hours

Comment (Rohesie) [reply code/game/objects/effects/mines.dm:86]: I mean you should use `smart_mine` as a variable name instead of `smartmine`, as it's more readable.

**Pull request closed without merging**